File: srcJava/com/rti/perftest/ddsimpl/DataTypeKeyedHelper.java
/*
 * (c) 2005-2017  Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

package com.rti.perftest.ddsimpl;

import java.util.List;
import com.rti.dds.topic.TypeSupportImpl;
import com.rti.perftest.TestMessage;
import com.rti.perftest.gen.TestDataKeyed_t;
import com.rti.perftest.gen.TestDataKeyed_tSeq;
import com.rti.perftest.gen.TestDataKeyed_tTypeSupport;
import com.rti.perftest.harness.PerfTest;


public class DataTypeKeyedHelper implements TypeHelper<TestDataKeyed_t> {

    public DataTypeKeyedHelper(int maxPerftestSampleSize) {
        _myData = new TestDataKeyed_t();
        _maxPerftestSampleSize = maxPerftestSampleSize;
    }

    public DataTypeKeyedHelper(TestDataKeyed_t myData, int maxPerftestSampleSize) {
        _myData = myData;
        _maxPerftestSampleSize = maxPerftestSampleSize;
    }

    public void fillKey(int value) {
        _myData.key[0] = (byte) (value);
        _myData.key[1] = (byte) (value >>> 8);
        _myData.key[2] = (byte) (value >>> 16);
        _myData.key[3] = (byte) (value >>> 24);
    }

    public void copyFromMessage(TestMessage message) {

        _myData.entity_id = message.entity_id;
        _myData.seq_num = message.seq_num;
        _myData.timestamp_sec = message.timestamp_sec;
        _myData.timestamp_usec = message.timestamp_usec;
        _myData.latency_ping = message.latency_ping;
        _myData.bin_data.loan(message.data, message.size);

    }

    @SuppressWarnings("rawtypes")
    public TestMessage copyFromSeqToMessage(List dataSeq, int index) {

        TestDataKeyed_t msg = (TestDataKeyed_t)((List)dataSeq).get(index);
        TestMessage message = new TestMessage();

        message.entity_id = msg.entity_id;
        message.seq_num = msg.seq_num;
        message.timestamp_sec = msg.timestamp_sec;
        message.timestamp_usec = msg.timestamp_usec;
        message.latency_ping = msg.latency_ping;
        message.size = msg.bin_data.size();
        message.data = (byte[]) msg.bin_data.getPrimitiveArray();

        return message;
    }

    public TestDataKeyed_t getData() {
        return _myData;
    }

    public void setBinDataMax(int newSize) {
        _myData.bin_data.setMaximum(newSize);
        _myData.bin_data.setSize(newSize);
    }

    public void bindataUnloan() {
        _myData.bin_data.unloan();
    }

    public TypeSupportImpl getTypeSupport() {
        return TestDataKeyed_tTypeSupport.getInstance();
    }

    public TypeHelper<TestDataKeyed_t> clone(){
        return new DataTypeKeyedHelper(_myData, _maxPerftestSampleSize);
    }

    @SuppressWarnings("rawtypes")
    public List createSequence() {
        return new TestDataKeyed_tSeq();
    }

    @SuppressWarnings("rawtypes")
    public int getMaxPerftestSampleSize() {
        return _maxPerftestSampleSize;
    }

    private int _maxPerftestSampleSize = PerfTest.getMaxPerftestSampleSizeJava();
    private TestDataKeyed_t _myData;

}

File: srcCs/ConnextDDS/TypeHelpers/DataTypeKeyedHelper.cs
/*
 * (c) 2005-2021 Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

using System.Linq;

namespace PerformanceTest
{
    public class DataTypeKeyedHelper : ITypeHelper<TestDataKeyed_t>
    {
        protected TestDataKeyed_t dataSample;
        protected TestMessage message = new TestMessage();
        protected ulong maxPerftestSampleSize = Perftest.GetMaxPerftestSampleSize();

        public DataTypeKeyedHelper(ulong maxPerftestSampleSize)
        {
            dataSample = new TestDataKeyed_t();
            this.maxPerftestSampleSize = maxPerftestSampleSize;
        }

        public DataTypeKeyedHelper(TestDataKeyed_t dataSample, ulong maxPerftestSampleSize)
        {
            this.dataSample = dataSample;
            this.maxPerftestSampleSize = maxPerftestSampleSize;
        }

        public TestDataKeyed_t MessageToSample(TestMessage message, int keyValue)
        {
            dataSample.key[0] = (byte)(keyValue);
            dataSample.key[1] = (byte)(keyValue >> 8);
            dataSample.key[2] = (byte)(keyValue >> 16);
            dataSample.key[3] = (byte)(keyValue >> 24);
            dataSample.entity_id = message.entityId;
            dataSample.seq_num = message.seqNum;
            dataSample.timestamp_sec = message.timestampSec;
            dataSample.timestamp_usec = message.timestampUsec;
            dataSample.latency_ping = message.latencyPing;

            DataHelper.PopulateData(dataSample.bin_data, message.data);

            return dataSample;
        }

        public TestMessage SampleToMessage(TestDataKeyed_t data_sample)
        {
            message.key = data_sample.key;
            message.entityId = data_sample.entity_id;
            message.seqNum = data_sample.seq_num;
            message.timestampSec = data_sample.timestamp_sec;
            message.timestampUsec = data_sample.timestamp_usec;
            message.latencyPing = data_sample.latency_ping;
            message.Size = data_sample.bin_data.Count;

            return message;
        }

        public ITypeHelper<TestDataKeyed_t> Clone()
        {
            return new DataTypeKeyedHelper(dataSample, maxPerftestSampleSize);
        }

        public long GetSerializedOverheadSize()
        {
            using (var serializer = TestDataKeyed_tSupport.Instance.CreateSerializer())
            {
                return serializer.GetSerializedSampleSize(dataSample) - Perftest.CDR_ENCAPSULATION_HEADER_SIZE;
            }
        }
    }
}